cellmlmanip/mathml2sympy/transpiler.py

```diff
@@ -0,0 +1,440 @@
+"""
+Parses Content MathML and returns equivalent SymPy expressions
+
+Content Markup specification: https://www.w3.org/TR/MathML2/chapter4.html
+
+TODO: Validate MathML input
+"""
+from xml.dom import minidom, Node
+
+import sympy
+
+
+def parse_string(xml_string):
+    """
+    Reads MathML content from a string and returns equivalent SymPy expressions
+    """
+    dom = minidom.parseString(xml_string)
+    return parse_dom(dom.childNodes[0])
+
+
+def parse_dom(math_dom_element):
+    """
+    Accepts a <math> node of DOM structure and returns equivalent SymPy expressions.
+    Note: math_dom_element must point the <math> XmlNode, not the root XmlDocument
+
+    TODO: adding cellml namespace here, but really should be provided by caller
+
+    :param math_dom_element: <math> XmlNode object of a MathML DOM structure
+    :return: List of SymPy expression(s)
+    """
+    math_dom_element.setAttributeNS('cellml', 'xmlns:cellml', 'http://www.cellml.org/cellml/1.0#')
+    return transpile(math_dom_element)
+
+
+def transpile(xml_node):
+    """
+    Descends the given MathML element node and calls the corresponding handler for child elements.
+    Returns the SymPy expression of node
+    :param xml_node: a DOM element of parsed MathML
+    :return: a list of SymPy expressions
+    """
+    # Collect the parsed expression(s) (i.e. SymPy output) into list
+    sympy_expressions = []
+
+    # For each child element of this DOM node
+    for child_node in xml_node.childNodes:
+        # If this is a Text node (no child nodes)
+        if child_node.nodeType == Node.TEXT_NODE:
+            # We should be handling text between tags explicitly in the handler
+            # (see cn_handler for an example), show a message
+            text = child_node.data.strip()
+            if text:
+                print('Hit text node with text "' + text + '"')
+        elif child_node.nodeType == child_node.ELEMENT_NODE:
+            # Call the appropriate MathML handler function for this tag
+            name = child_node.tagName
+            if name in HANDLERS:
+                # If this tag element itself has children
+                if child_node.childNodes:
+                    # We want to pass the node to the handler, and it will deal with children
+                    sympy_expressions.append(HANDLERS[name](child_node))
+                else:
+                    # This tag has no children
+                    sympy_expressions.append(HANDLERS[name]())
+            else:
+                # MathML handler function not found for this tag!
+                raise NotImplementedError('No handler for element <%s>' % child_node.tagName)
+        else:
+            raise NotImplementedError('Unknown node type %d' % child_node.nodeType)
```

Comment:
We should probably silently ignore comments, at least. Processing instructions too, I guess.

Reply:
Done